pkg/storage/storage_provisioning.go

```diff
@@ -0,0 +1,200 @@
+	if persistent {
+		removeOnDelete = "false"
+	} else {
+		removeOnDelete = "true"
+	}
+	existing.Labels["remove-on-delete"] = removeOnDelete
+	return r.Client.Update(context.Background(), existing)
+}
+
+// GetDatasetStatus returns status of an existing Dataset resource.
+func (r *ProvisionImpl) GetDatasetStatus(ref *types.NamespacedName) (*ProvisionedStorageStatus, error) {
+	dataset := &comv1alpha1.Dataset{}
+	if err := r.Client.Get(context.Background(), types.NamespacedName{Name: ref.Name, Namespace: ref.Namespace}, dataset); err != nil {
+		return nil, err
+	}
+	return &ProvisionedStorageStatus{Provisioned: dataset.Status.Provision.Status == "OK", ErrorMsg: dataset.Status.Provision.Info}, nil
+}
+
+// DeleteDataset deletes the existing Dataset resource
+func (r *ProvisionImpl) DeleteDataset(ref *types.NamespacedName) error {
+	dataset := &comv1alpha1.Dataset{}
+	if err := r.Client.Get(context.Background(), types.NamespacedName{Name: ref.Name, Namespace: ref.Namespace}, dataset); err != nil {
+		return err
+	}
+	return r.Client.Delete(context.Background(), dataset)
+}
+
+// ProvisionTest is an implementation of ProvisionInterface used for testing
+type ProvisionTest struct {
+	datasets []*ProvisionedBucket
+}
+
+// NewProvisionTest constructs a new ProvisionTest object
+func NewProvisionTest() *ProvisionTest {
+	return &ProvisionTest{
+		datasets: []*ProvisionedBucket{},
+	}
+}
+
+// CreateDataset generates a new dataset
+func (r *ProvisionTest) CreateDataset(ref *types.NamespacedName, dataset *ProvisionedBucket, owner *types.NamespacedName) error {
+	for i, d := range r.datasets {
+		if d.Name == dataset.Name {
+			r.datasets[i] = dataset
+			return nil
+		}
+	}
+	r.datasets = append(r.datasets, dataset)
+	return nil
+}
+
+// SetPersistent does nothing for the testing implementation except for verifying that the dataset exists
+func (r *ProvisionTest) SetPersistent(ref *types.NamespacedName, persistent bool) error {
+	for _, d := range r.datasets {
+		if d.Name == ref.Name {
+			return nil
+		}
+	}
+	return errors.New("Could not find a dataset: " + ref.Name)
+}
+
+// GetDatasetStatus returns status of an existing Dataset resource.
+func (r *ProvisionTest) GetDatasetStatus(ref *types.NamespacedName) (*ProvisionedStorageStatus, error) {
+	for _, d := range r.datasets {
+		if d.Name == ref.Name {
+			return &ProvisionedStorageStatus{Provisioned: true}, nil
+		}
+	}
+	return nil, errors.New("Could not find a dataset: " + ref.Name)
+}
+
+// DeleteDataset removes an existing dataset
+func (r *ProvisionTest) DeleteDataset(ref *types.NamespacedName) error {
+	newDatasets := []*ProvisionedBucket{}
+	found := false
+	message := "Datasets:\n"
+	for _, d := range r.datasets {
+		message += " " + d.Name + "\n"
+		if d.Name == ref.Name {
+			found = true
+		} else {
+			newDatasets = append(newDatasets, d)
+		}
+	}
+	if found {
+		r.datasets = newDatasets
+		return nil
+	}
+	return errors.New("Could not delete a dataset: " + ref.Name + "\n" + message)
```

Comment:
what does "message" here means? the list of current datasets? should we reflect it to make it clear?

Reply:
Modified.